core/src/common/empty_block/inherents/custom_idps/para_parachain.rs

```diff
@@ -0,0 +1,130 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Inherent data provider for the [cumulus parachin inherents](https://github.com/paritytech/polkadot-sdk/blob/master/cumulus/primitives/parachain-inherent/src/lib.rs)
+//! for empty block production on top of an existing externalities.
+
+use parity_scale_codec::{Decode, Encode};
+use polkadot_primitives::{BlockNumber, HeadData};
+use sp_consensus_babe::SlotDuration;
+use sp_core::twox_128;
+use sp_inherents::InherentIdentifier;
+use sp_runtime::traits::{Block as BlockT, HashingFor, NumberFor};
+use sp_state_machine::TestExternalities;
+
+/// ext cannot be part of the InherentDataProvider (thread safety), so we need to make storage
```

Comment:
Because its not `Send`? There is [`spawn_local`](https://docs.rs/tokio/latest/tokio/task/fn.spawn_local.html) to spawn non `Send` closures, maybe that works.

Reply:
It's because some nested `RefCell` in it is not `Sync`.

I found a way to fix this by wrapping `TestExternalities` in `Arc<Mutex<...`, it's a bit uglier but avoids this issue.